File: Golang/valid_parentheses/20.go
package validparentheses

func isValid(s string) bool {
	if len(s)%2 != 0 {
		return false
	}

	stack := []rune{}

	for _, char := range s {
		if char == '(' || char == '[' || char == '{' {
			stack = append(stack, char)
		} else {
			if len(stack) == 0 {
				return false
			} else if (char == ')' && stack[len(stack)-1] == '(') || (char == ']' && stack[len(stack)-1] == '[') || (char == '}' && stack[len(stack)-1] == '{') {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}

	return len(stack) == 0
}

File: C++/src/20.cpp
#include <solution_tmpl.h>

bool Solution::isValid(string s) {
  int length = s.length();
  if (length % 2 != 0) {
    return false;
  }

  vector<char> stack = {};
  for (int i = 0; i < length; i++) {
    if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
      stack.push_back(s[i]);
    } else {
      if (stack.empty()) {
        return false;
      }

      if ((s[i] == ')' && stack.back() == '(') ||
          (s[i] == '}' && stack.back() == '{') ||
          (s[i] == ']' && stack.back() == '[')) {
        stack.pop_back();
      } else {
        return false;
      }
    }
  }

  return stack.empty();
}